src/sentencepiece_processor.cc

```diff
@@ -594,13 +608,51 @@ util::Status SentencePieceProcessor::Decode(
     if (!IsByte(sp.id())) {
       RETURN_IF_ERROR(ProcessBytePieces(byte_start, i));
       byte_start = i + 1;
-      SetSurface(i, DecodeSentencePiece(sp.piece(), sp.id(), text->empty()));
+      bool is_eos_space = i == spt->pieces_size() - 1;
+      SetSurface(i, DecodeSentencePiece(sp.piece(), sp.id(), text->empty(), is_eos_space));
     }
   }
   RETURN_IF_ERROR(ProcessBytePieces(byte_start, spt->pieces_size()));
 
   if (denormalizer_) {
```

Comment:
Please document the code inside this block a bit more. Maybe even consider extracting it (or a part of it) into a subroutine if appropriate.